### ISSUE TITLE
ResultSet.getBigDecimal returns 0 instead of null for null value

### ISSUE THREAD

**Issue by eirikbakke** [xerial/sqlite-jdbc#801]
**Describe the bug**
The ResultSet.getBigDecimal(String) method returns 0 instead of null when a null value is retrieved. This is incorrect; the getBigDecimal Javadoc states "if the value is SQL NULL, the value returned is null in the Java programming language".

This bug occurs in Xerial JDBC 3.39.2.0 but not on 3.36.0.3. The change in question might be https://github.com/xerial/sqlite-jdbc/pull/666

**To Reproduce**

```java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

public class SQLiteGetBigDecimalNullBug {
  public static void main(String[] args) throws SQLException {
    try (Connection connection = DriverManager.getConnection("jdbc:sqlite::memory:")) {
      Statement statement = connection.createStatement();
      statement.executeUpdate("create table testtab (id string, content integer)");
      statement.executeUpdate("insert into testtab values('A', 52)");
      statement.executeUpdate("insert into testtab values('B', null)");
      ResultSet rs = statement.executeQuery("select id, content from testtab");
      while (rs.next()) {
        System.out.println("id = " + rs.getString("id"));
        // For row B, prints null on Xerial JDBC 3.36.0.3 (correct), but 0 on 3.39.2.0 (incorrect).
        System.out.println("content = " + rs.getBigDecimal("content"));
      }
    }
  }
}
```

**Expected behavior**
For row B, code should print "content = null".

**Logs**
N/A

**Environment (please complete the following information):**
 - OS: Windows 11
 - CPU architecture: x86_64
 - sqlite-jdbc version: 3.39.2.0 and 3.39.3.0

**Additional context**
N/A

**Comment by gotson** [comment]
i think the problem comes from the column definition in your create table statement :`create table testtab (id string, content integer)`

If you change this to:
- `create table testtab (id string, content real)`
- `create table testtab (id string, content bigint)`

Then it works as expected.

The problem lies in how the driver tries to guess the type, as SQLite is loosely typed, it will check the declared type, and the actual storage type.

I think the problem [is here](https://github.com/xerial/sqlite-jdbc/blob/7912b6b5620f3c70fd8c42d3a9ce676048accd12/src/main/java/org/sqlite/jdbc3/JDBC3ResultSet.java#L631-L631), which ignores the actual type (NULL), and returns INTEGER instead.

Basically we are losing the actual storage class, and this happens in other places too. SQLite recommends checking the actual type before trying to retrieve the value.

We shouldn't need to perform a `getString` to check whether a value is `null`.

**Comment by eirikbakke** [comment]
> SQLite recommends checking the actual type before trying to retrieve the value.

Looking at the source code for getBigDecimal, perhaps safeGetColumnType should be used instead of getColumnType? I.e. it seems the first one get the actual type of the individual value, while the latter gets the approximated type for the entire column?

**Comment by gotson** [comment]
> > SQLite recommends checking the actual type before trying to retrieve the value.
> 
> Looking at the source code for getBigDecimal, perhaps safeGetColumnType should be used instead of getColumnType? I.e. it seems the first one gets the actual type of the individual value, while the latter gets the approximated type for the entire column?

I just fixed this and added some tests.

Basically the function was reusing `getColumnType` which as you found out is inferring the type depending on the declared type and the actual type.

I added a check at the beginning to check the actual type, and return `null` if the SQLite storage class is null.

**Comment by eirikbakke** [comment]
Thanks for the fix!

Looking at the commit... should getColumnType (as opposed to safeGetColumnType) be used at all, though? Couldn't safeGetColumnType be used to decide whether to use getDouble (in the SQLITE_FLOAT case) or getString?

**Comment by gotson** [comment]
> Looking at the commit... should getColumnType (as opposed to safeGetColumnType) be used at all, though? Couldn't safeGetColumnType be used to decide whether to use getDouble (in the SQLITE_FLOAT case) or getString?

yes, because we try to provide a value that is in line with the declared type of the column

**Comment by eirikbakke** [comment]
> yes, because we try to provide a value that is in line with the declared type of the column

Is this a general policy, or specific to getBigDecimal? I don't see any other uses of getColumnType or getColumnTypeName, except in isSigned().

If the goal is to preserve the value's original precision, using BigDecimal.valueOf(safeGetDoubleCol) for SQLITE_FLOAT (from safeGetColumnType()) and BigDecimal.valueOf(safeGetColumnText()) for SQLITE_TEXT would seem to make the most sense. In either case the client has specified that they want a decimal type by calling getBigDecimal.

**Comment by gotson** [comment]
After checking the code and the previous PR, i agree with you. Will do the changes.

**Comment by eirikbakke** [comment]
The latest version makes a lot of sense to me! Thank you for fixing this, and for this great library in general!
